larpandoracontent/LArHelpers/LArHierarchyHelper.h

```diff
@@ -0,0 +1,629 @@
+         *
+         *          If only folding back to leading shower particles, the hierarchy will largely mirror the standard PFO hierarchy, but,
+         *          when a shower particle is reached (based on the track/shower characterisation), this particle and all daughter particles
+         *          will be represented by a single leaf node.
+         *
+         *          If folding back to both primary and leading shower particles the hierarchy will again be rather flat, but in this case,
+         *          if a primary track-like particle has a downstream shower particle then all downstream particles above the shower-like
+         *          particle will be folded into the primary node, but a new, daughter leaf node will be created for the shower-like
+         *          particle and all of its daughters, and a parent-child relationship will be formed between the primary node and shower
+         *          node.
+         *
+         *  @param  pfoList The list of PFOs with which to fill the hierarchy
+         *  @param  foldToPrimaries Whether or not to fold daughter particles back to their primary particle
+         *  @param  foldToLeadingShowers Whether or not to fold daughter particles back to their leading shower particle
+         */
+        void FillHierarchy(const pandora::PfoList &pfoList, const bool foldToPrimaries, const bool foldToLeadingShowers);
+
+        // Note - likely want to change this to retrieve the neutrino, the test beam particle, all primaries, all cosmics etc
+        /**
+         *  @brief  Retrieve the root nodes in this hierarchy
+         *
+         *  @return The root nodes in this hierarchy
+         */
+        const NodeVector& GetRootNodes() const { return m_rootNodes; }
+
+        /**
+         *  @brief  Retrieve a flat vector of the ndoes in the hierarchy
+         *
+         *  @param  nodeVector The output vector for the nodes in the hierarchy in breadth first order
+         */
+        void GetFlattenedNodes(NodeVector &nodeVector) const;
+
+        /**
+         *  @brief  Produce a string representation of the hierarchy
+         *
+         *  @return The string representation of the hierarchy
+         */
+        const std::string ToString() const;
+
+    private:
+        NodeVector m_rootNodes;
+        const pandora::ParticleFlowObject *m_pNeutrino;
+    };
+
+    class MCMatches
+    {
+    public:
+        /**
+         *  @brief  Constructor
+         *
+         *  @param  pMC The MCParticle being matched
+         */
+        MCMatches(const MCHierarchy::Node *pMC);
+
+        virtual ~MCMatches() = default;
+
+        // Might be better to pass in the shared hits themselves
+        /**
+         *  @brief  Add a reconstructed node as a match for this MC node
+         *
+         *  @param  pReco The reconstructed node that matches this MC node
+         */
+        void AddRecoMatch(const RecoHierarchy::Node *pReco, const int nSharedHits);
```

Comment:
nSharedHits missing description... (unless you plan to change the argument as per your comment above)

Reply:
Ah yes, I did omit for that reason, but actually, I'm currently planning to stick with the count, so will need to document this.